src/occupation.jl

```diff
@@ -3,67 +3,39 @@
 import Roots
 
 """
-Find Fermi level and occupation for the given parameters, assuming a band gap
-and zero temperature.
+Find the Fermi level, given a `basis`, SCF band `energies` and corresponding Bloch
+one-particle wave function `Psi`.
 """
-function find_occupation_gap_zero_temperature(basis, energies, Psi)
-    @assert basis.model.spin_polarisation in (:none, :spinless)
-    n_bands = size(Psi[1], 2)
-    n_electrons = basis.model.n_electrons
-    T = eltype(basis.kpoints[1].coordinate)
-    @assert basis.model.temperature == 0.0
-
-    filled_occ = filled_occupation(basis.model)
-    n_fill = div(n_electrons, filled_occ)
-    @assert filled_occ * n_fill == n_electrons
-    @assert n_bands ≥ n_fill
-
-    # We need to fill n_fill states with occupation filled_occ
-    # Find HOMO and LUMO
-    HOMO = -Inf # highest occupied energy state
-    LUMO = Inf  # lowest unoccupied energy state
-    occupation = similar(basis.kpoints, Vector{T})
-    for ik in 1:length(occupation)
-        occupation[ik] = zeros(T, n_bands)
-        occupation[ik][1:n_fill] .= filled_occ
-        HOMO = max(HOMO, energies[ik][n_fill])
-        if n_fill < n_bands
-            LUMO = min(LUMO, energies[ik][n_fill + 1])
-        end
-    end
-    @assert sum(basis.kweights .* sum.(occupation)) ≈ n_electrons
-
-    # Put Fermi level slightly above HOMO energy, to ensure that HOMO < εF
-    εF = nextfloat(HOMO)
-    if εF > LUMO
-        @warn("`find_occupation_gap_zero_temperature` assumes an insulator, but the " *
-              "system seems metallic. Try specifying a temperature and a smearing function.",
-              HOMO, LUMO, maxlog=5)
-    end
-
-    εF, occupation
+function find_fermi_level(basis, energies)
+    εF, _ = find_occupation(basis, energies)
+    εF
 end
 
 """
-Find the Fermi level and occupation for the given parameters, assuming no band gap
-(i.e. a metal).
+Find the Fermi level and occupation for the given parameters.
 """
-function find_occupation_fermi_metal(basis, energies, Psi)
+function find_occupation(basis::PlaneWaveBasis{T}, energies) where {T}
     @assert basis.model.spin_polarisation in (:none, :spinless)
     model = basis.model
-    n_bands = size(Psi[1], 2)
     n_electrons = model.n_electrons
     temperature = model.temperature
     smearing = model.smearing
 
+    # Avoid numerical issues by imposing a step function for zero temp
+    temperature == 0 && (smearing(x) = x ≤ 0 ? 1 : 0)
     @assert smearing !== nothing
 
     filled_occ = filled_occupation(model)
 
-    @assert filled_occ*n_bands ≥ n_electrons
+    # assume that we can get the required number of electrons by filling every state
+    @assert filled_occ*sum(basis.kweights .* length.(energies)) ≥ n_electrons - sqrt(eps(T))
 
-    # Avoid numerical issues by imposing a step function for zero temp
-    temperature == 0 && (smearing(x) = x ≤ 0 ? 1 : 0)
+    # early exit when all bands have to be filled; avoids numerical problems
```

Comment:
Just out of interest: Do you not expect problems if the lowermost levels need to be fully filled (e.g. the insulator case)?

Reply:
Not really, why? This is the standard way to do things for insulators.

Reply:
What I meant is if you have an insulator and `n_bands > n_electrons // filled_occ`

Reply:
I simplified and commented that branch. The only trouble is when there is no wiggle room, so that bracketing doesn't work

Reply:
Perfect. That was the part I missed so far.

Reply:
Yeah, these rounds of review are pretty effective at forcing me to be less lazy :-)